# Проект по SQL

# Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Моя компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Моя первая задача как аналитика — проанализировать базу данных.

В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

# Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

# Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

# Цель исследования

Проанализировать базу данных при помощи SQL-запросов и сформулировать ценностное предложение для нового продукта.

Подключимся в базе данных:

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# Таблица 'books'

In [2]:
query = '''
            SELECT * FROM books
        '''
books = pd.io.sql.read_sql(query, con = engine)

books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Узнаем число записей в таблице:

In [3]:
request = '''
             SELECT COUNT(*)FROM books
          '''
pd.io.sql.read_sql(request, con = engine)

count
0   1000

В таблице 'books' 1000 записей.

# Таблица 'authors'

In [4]:
query = '''
            SELECT * FROM authors
        '''
authors = pd.io.sql.read_sql(query, con = engine)

authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Узнаем число записей в таблице:

In [5]:
request = '''
             SELECT COUNT(*)FROM authors
          '''
pd.io.sql.read_sql(request, con = engine)

count
0    636

В таблице 'authors' 636 записей.

# Таблица 'publishers'

In [6]:
query = '''
            SELECT * FROM publishers
        '''
publishers = pd.io.sql.read_sql(query, con = engine)

publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Узнаем число записей в таблице:

In [7]:
request = '''
             SELECT COUNT(*)FROM publishers
          '''
pd.io.sql.read_sql(request, con = engine)

count
0    340

В таблице 'publishers' 340 записей.

# Таблица 'ratings'

In [8]:
query = '''
            SELECT * FROM ratings
        '''
ratings = pd.io.sql.read_sql(query, con = engine)

ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Узнаем число записей в таблице:

In [9]:
request = '''
             SELECT COUNT(*)FROM ratings
          '''
pd.io.sql.read_sql(request, con = engine)

count
0   6456

В таблице 'ratings' 6456 записей.

# Таблица 'reviews'

In [10]:
query = '''
            SELECT * FROM reviews
        '''
reviews = pd.io.sql.read_sql(query, con = engine)

reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Узнаем число записей в таблице:

In [11]:
request = '''
             SELECT COUNT(*)FROM reviews
          '''
pd.io.sql.read_sql(request, con = engine)

count
0   2793

В таблице 'ratings' 2793 записей.

# Решение задач

1. Посчитайте, сколько книг вышло после 1 января 2000 года:

In [12]:
query = '''

SELECT COUNT(book_id)
FROM books 
WHERE publication_date > '2000-01-01'

'''

query_1 = pd.io.sql.read_sql(query, con = engine) 
query_1

count
0    819

После 1 января 2000 года вышло 819 книг

2. Для каждой книги посчитайте количество обзоров и среднюю оценку:

In [13]:
query = '''

SELECT b.title,
       COUNT (DISTINCT rw.review_id), ROUND (avg(r.rating), 2) as R
FROM REVIEWS AS rw
RIGHT OUTER JOIN BOOKS AS b ON rw.book_id = b.book_id
LEFT OUTER JOIN RATINGS AS r ON b.book_id = r.book_id
GROUP BY b.book_id
ORDER BY R DESC

'''

query_2 = pd.io.sql.read_sql(query, con = engine)
query_2.columns = ['Название книги', 'Количество обзоров', 'Средняя оценка']
query_2

Название книги  Количество обзоров  \
0        Arrows of the Queen (Heralds of Valdemar  #1)                   2   
1    The Walking Dead  Book One (The Walking Dead #...                   2   
2                                      Light in August                   2   
3    Wherever You Go  There You Are: Mindfulness Me...                   2   
4    Captivating: Unveiling the Mystery of a Woman'...                   2   
..                                                 ...                 ...   
995  The World Is Flat: A Brief History of the Twen...                   3   
996                  His Excellency: George Washington                   2   
997                                      Drowning Ruth                   3   
998                                              Junky                   2   
999                               Harvesting the Heart                   2   

     Средняя оценка  
0              5.00  
1              5.00  
2              5.00  
3              5.00  
4              5.00  
..              ...  
995            2.25  
996            2.00  
997            2.00  
998            2.00  
999            1.50  

[1000 rows x 3 columns]

Количество обзоров и средняя оценка рассчитаны по каждой книге и добавлены в таблицу в новые столбцы

3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц:

In [14]:
query = '''

SELECT p.publisher,
       count(b.book_id)
FROM publishers AS p
JOIN BOOKS AS b ON p.publisher_id=b.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher_id
ORDER BY count(b.book_id) DESC
LIMIT 1

'''

query_3 = pd.io.sql.read_sql(query, con = engine)
query_3.columns = ['Название издательства', 'Количество книг']
query_3

Название издательства  Количество книг
0         Penguin Books               42

Издательство, которое выпустило наибольшее число книг толще 50 страниц, называется "Penguin Books"

4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками:

In [15]:
query = '''
SELECT author,
       ROUND(AVG(rating), 2) as avg_rating
FROM authors AS a
INNER JOIN books AS b ON b.author_id=a.author_id
INNER JOIN ratings AS r ON r.book_id=b.book_id
WHERE b.book_id IN (SELECT  book_id
                            FROM ratings
                            GROUP BY book_id
                            HAVING COUNT(rating) > 50)
GROUP BY author
ORDER BY AVG(rating) DESC
LIMIT 1

'''

query_4 = pd.io.sql.read_sql(query, con = engine)
query_4

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

Автора с самой высокой средней оценкой книг (среди книг от с числом оценок от 50 и более) зовут J.K. Rowling/Mary GrandPré

5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок:

In [16]:
query = '''

WITH A AS
  (SELECT username,
          count(review_id)
   FROM reviews
   WHERE username in
       (SELECT username
        FROM ratings
        GROUP BY username
        HAVING count(rating_id) > 48)
   GROUP BY username)
SELECT round (SUM (COUNT) / COUNT(COUNT), 2) AS AVG
FROM A

'''

query_5 = pd.io.sql.read_sql(query, con = engine)
query_5

avg
0  24.0

Пользователи, оставившие более 48 оценок, в среднем сделали 24 обзора

# Выводы

Была проанализирована базу данных сервиса чтения книг. Она включала такие таблицы:
* books
* authors
* publishers
* ratings
* reviews

В результате анализа мы выяснили следующее:
* После 1 января 2000 года вышло 819 книг.
* Издательство, которое выпустило наибольшее число книг толще 50 страниц, называется "Penguin Books".
* Автора с самой высокой средней оценкой книг (среди книг от с числом оценок от 50 и более) зовут J.K. Rowling/Mary GrandPré.
* Пользователи, оставившие более 48 оценок, в среднем сделали 24 обзора.

Также были рассчитаны и добавлены в таблицу количество обзоров и средние оценки по каждой книге.

# Рекомендации

Для получения хорошей прибыли в ассортименте сервиса нужно в первую очередь:
* иметь в продаже книги с высокой средней оценкой и большим количеством обзоров, как наиболее востребованных потенциальными покупателями;
* иметь в продаже максимально возможный ассортимент книг авторов с высокой средней оценкой и большим количеством отзывов, как наиболее востребованных потенциальными покупателями;
* обратить внимание на книги, выпущенные издательством "Penguin Books", как наиболее плодотворное в плане выпуска книг толще 50 страниц.